PLUGIN-10. Jira Batch Source

aonischuk opened the pull request:
I split the PR into two commits so it's simpler to review.

Design doc:
https://wiki.cask.co/display/CE/Jira+Source

The plugin fetches the issues from jira in a parallel fashion. Jira java API is used.
User can specify three filters for issues:
- by jira filter id
- by JQL query
- by set of plugin configs (fixVersion, assignee, updateDate etc.)

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+Description
+-----------
+The plugin fetches issues from JIRA in a parallel fashion.
+JQL or existing JIRA filters or custom filters via configurations can be used 
```
JQL, existing JIRA filters, or custom filters can be used in order to read a subset of issues.

aonischuk replied:
Done

albertshau commented on widgets/Jira-batchsource.json:
```diff
@@ -0,0 +1,593 @@
+    {
+      "name": "Filter Mode 'JQL'",
+      "condition": {
+        "expression": "filterMode == 'JQL'"
```
avoid using expressions unless it can't be done otherwise. See https://github.com/data-integrations/http/blob/develop/widgets/HTTP-batchsource.json#L673 for equality.

aonischuk replied:
Done

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+
+**Jira URL:** URL of JIRA instance.
+
+**Filter Mode:** Mode which specifies which issues to fetch.
```
I think the filter makes more sense to put in 'Advanced', since it's not required and some users may not know how to configure it.

aonischuk replied:
Done

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+
+***Basic Filter Mode***
+
+Enables to specify basic filters using plugin properties.
```
I assume if multiple properties are given, they all need to match? i.e. project == 'xyz' and status = 'open'? Assuming that's the case:

Simple value filters. If multiple properties are given, they must all match in order for the issue to be read. For example, if Projects is set to 'CDAP' and 'Statuses' is set to 'Open', the issue must belong to project 'CDAP' and have an 'Open' status in order.

aonischuk replied:
That is true they all need to match. The query will look like this:
`project IN ('CDAP') AND fixVersion IN ('6.1.0', '6.2.0') AND status IN ('Open')`

albertshau replied:
Please update the doc accordingly.

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+
+Related configurations:
+
+**Projects**: List of jira projects used for filtering.
```
can remove 'used for filtering' from these descriptions.

List of jira projects to match.

aonischuk replied:
Done

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+
+**Labels**: List of issue labels used for filtering.
+
+**Last Update Start Date**: Minimal update date for the issues fetched.
```
Earliest update date to include.

aonischuk replied:
Done

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+Valid formats include: 'yyyy/MM/dd HH:mm', 'yyyy-MM-dd HH:mm', 'yyyy/MM/dd', 
+'yyyy-MM-dd', or a period format e.g. '-5d', '4w 2d'.
+
+**Last Update End Date**: Maximum update date for the issues fetched.
```
Latest update date to include.

aonischuk replied:
Done

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+
+**Last Update Start Date**: Minimal update date for the issues fetched.
+Valid formats include: 'yyyy/MM/dd HH:mm', 'yyyy-MM-dd HH:mm', 'yyyy/MM/dd', 
+'yyyy-MM-dd', or a period format e.g. '-5d', '4w 2d'.
```
Is this period format something jira supports?

aonischuk replied:
Yes this is something supported by jira JQL. And actually this text is just a copy from their docs.

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -0,0 +1,87 @@
+
+### Advanced
+
+**Max Split Size:** Maximum number of issue which are fetched using a single request and 
```
Maximum number of issues to fetch in a single request. A value of zero means there is no limit.

aonischuk replied:
Done

aonischuk commented:
@albertshau thank you for your review. I have addressed all the comments. Can you please look again?

albertshau commented on docs/Jira-batchsource.md:
```diff
@@ -30,32 +23,32 @@ should only conform to one of the values specified.
 
 Related configurations:
```
this section also need to be moved down, under the Filter Mode

albertshau commented on widgets/Jira-batchsource.json:
```diff
@@ -17,19 +17,6 @@
-        },
         {
           "widget-type": "csv",
           "label": "Projects",
```
also move these down, under the Filter Mode

albertshau commented:
minor comments, please fix and squash

aonischuk committed:
PLUGIN-10. Jira Batch Source

aonischuk commented:
@albertshau can you please merge? I have fixed the comments and squashed. Here's how the widget now looks like:

![Selection_041](https://user-images.githubusercontent.com/46598243/68055249-e2487880-fcf8-11e9-80f6-e6826b6cc266.png)
